### PR TITLE
Jmcneill/expression is true tweak

### PR DESCRIPTION
**Pull request by jpmmcneill**
This is a:
- [ ] bug fix PR with no breaking changes — please ensure the base branch is `main`
- [x] new functionality — please ensure the base branch is the latest `dev/` branch
- [ ] a breaking change — please ensure the base branch is the latest `dev/` branch

## Description & motivation
Adjust the way that the `expression_is_true` test gets evaluated (basically move it outside of a where clause).
Issue context: https://github.com/dbt-labs/dbt-utils/issues/490
This allows for use of window functions at a column level, and maybe a bunch of other stuff.
Tests have been added window functionality.

## Checklist
- [x] I have verified that these changes work locally on the following warehouses (Note: it's okay if you do not have access to all warehouses, this helps us understand what has been covered)
    - [ ] BigQuery
    - [x] Postgres
    - [ ] Redshift
    - [ ] Snowflake
- [x] I followed guidelines to ensure that my changes will work on "non-core" adapters by:
    - [ ] dispatching any new macro(s) so non-core adapters can also use them (e.g. [the `star()` source](https://github.com/fishtown-analytics/dbt-utils/blob/master/macros/sql/star.sql))
    - [ ] using the `limit_zero()` macro in place of the literal string: `limit 0`
    - [ ] using `dbt_utils.type_*` macros instead of explicit datatypes (e.g. `dbt_utils.type_timestamp()` instead of `TIMESTAMP`
- [ ] I have updated the README.md (if applicable)
- [x] I have added tests & descriptions to my models (and macros if applicable)
- [ ] I have added an entry to CHANGELOG.md

### CODE REVIEW

**Comment by jpmmcneill** [comment]
I accidentally based on main rather than next/minor. Happy to change this if needed.

I guess that is the reason for https://github.com/dbt-labs/dbt-utils/pull/507/commits/a7f4f51fe1d9883150f01f623e83f1ab197efcbb being included in the diff.

File: integration_tests/models/schema_tests/schema.yml [hunk 1]
```diff
@@ -31,6 +31,21 @@ models:
...
+          expression: max(col_c) over (partition by col_a) = min(col_c) over (partition by col_a)
+      - dbt_utils.expression_is_true:
+          expression: sum(col_b) over (partition by col_a) = 15
+          condition: col_a = 1
+    columns:
+      - name: col_c
+        tests:
+          - dbt_utils.expression_is_true:
+              expression: = sum(col_b) over (partition by col_a)
+
```
Comment (jpmmcneill) [inline integration_tests/models/schema_tests/schema.yml:48]: Test that window functions are behaving as expected

File: integration_tests/models/schema_tests/schema.yml [hunk 2]
```diff
@@ -13,7 +13,7 @@ models:
         tests:
           - dbt_utils.at_least_one
 
-  - name: data_test_expression_is_true
+  - name: data_test_expression_is_true_1
```
Comment (jpmmcneill) [inline integration_tests/models/schema_tests/schema.yml:16]: Migrate old seed to new one

File: macros/schema_tests/expression_is_true.sql [hunk 3]
```diff
@@ -7,16 +7,18 @@
...
 with meet_condition as (
-    select * from {{ model }} where {{ condition }}
+    select
+      *,
+      {% if column_name is none %}
+      {{ expression }}
+      {%- else %}
+      {{ column_name }} {{ expression }}
+      {%- endif %}
+      as _test_expression_passed
```
Comment (jpmmcneill) [inline macros/schema_tests/expression_is_true.sql:17]: The one thing I'm a bit nervous about here is a field name collision.

A possible solution is to grab the column names from the given model and throw a compiler message if _test_expression_passed is one of the field names.

However, that seems a bit contrived to me. Another possibility is just make the field name more constructed (ie. maybe encode the expression in it etc). Would like to hear peoples thoughts :)

**Review by jpmmcneill — COMMENTED** [review]
Adding notes for changes in comments

File: macros/schema_tests/expression_is_true.sql (hunk 3 above)
Comment (jpmmcneill) [reply macros/schema_tests/expression_is_true.sql:26]: I decided to catch this exception in: https://github.com/dbt-labs/dbt-utils/pull/507/commits/68dc34114e6636ee23495e72bed893c002a57ac4.

Let me know if its a bit expensive to use get columns in relation to catch this error, or if people should be left to fend for themselves when seeing "ambiguous column name" errors.

**Comment by jpmmcneill** [comment]
Hey @joellabes - you may have missed this. Nudging as it's ready for review :)

File: integration_tests/models/schema_tests/schema.yml (hunk 1 above)
Comment (joellabes) [reply integration_tests/models/schema_tests/schema.yml:48]: This is a great set of tests 😍

File: macros/schema_tests/expression_is_true.sql (hunk 3 above)
Comment (joellabes) [reply macros/schema_tests/expression_is_true.sql:26]: It's a fair concern! My gut feeling is that this is a reasonably obscure column name. Perhaps you could include the package name:
```suggestion
      as _dbt_utils_test_expression_passed
```

I don't think it's necessary to protect against the ambiguous column name, just selecting * should be OK.

File: macros/schema_tests/expression_is_true.sql [hunk 4]
```diff
@@ -6,17 +6,28 @@
...
 
-select
-    *
-from meet_condition
-{% if column_name is none %}
-where not({{ expression }})
-{%- else %}
-where not({{ column_name }} {{ expression }})
-{%- endif %}
+select * from meet_condition where not(_test_expression_passed)
```
Comment (joellabes) [inline macros/schema_tests/expression_is_true.sql:31]:
```suggestion
select * from meet_condition 
where not(_test_expression_passed)
```

File: integration_tests/models/schema_tests/schema.yml (hunk 2 above)
Comment (joellabes) [reply integration_tests/models/schema_tests/schema.yml:16]: Could you name the seeds with a bit more clarity? perhaps `data_test_expression_is_true` (i.e staying as-is) and `data_test_expression_is_true_window_functions` or something?

**Review by joellabes — CHANGES_REQUESTED** [review]
I like where this is going 😍 ! A couple of nitpicks and comments below.

A side note: Since this will change the shape of the results table, which _could_ be a breaking change for anyone who uses `--store-failures`, this will likely come out in 0.9.0 as opposed to a patch release. 

I'm OK with that though - we _could_ exclude the new column with `star`, but I think it's somewhere between `not harmful` and `helpful` to have the true/false result of the evaluation included.

File: integration_tests/models/schema_tests/schema.yml (hunk 2 above)
Comment (jpmmcneill) [reply integration_tests/models/schema_tests/schema.yml:16]: Happy to. If this is something that you don't like, this pattern of seed (`*_1.csv`, `*_2.csv` etc...) is present in a few other test files FYI

File: macros/schema_tests/expression_is_true.sql (hunk 3 above)
Comment (jpmmcneill) [reply macros/schema_tests/expression_is_true.sql:26]: Will do & i'll remove the get column in relation check.

**Comment by jpmmcneill** [comment]
@joellabes thanks for the review! I've updated with your comments in https://github.com/dbt-labs/dbt-utils/pull/507/commits/24ddca972e810fab8a041e01527de3adcbb4915c

> A side note: Since this will change the shape of the results table, which could be a breaking change for anyone who uses --store-failures, this will likely come out in 0.9.0 as opposed to a patch release.
I'm OK with that though - we could exclude the new column with star, but I think it's somewhere between not harmful and helpful to have the true/false result of the evaluation included.

On this, I'm certainly on team "it's helpful".

If i'm debugging a test, i'm always going to grab the compiled test code, run it and i'll want to see the result of some adjustments to pin down whats happening. In that case i'd like to see the result of the test to visualise how different things impact the result in something like a query editor - in my experience the fastest way to do this is with the test result as a column.

**Review by joellabes — APPROVED** [review]
Beautiful! Thanks @jpmmcneill - keep an eye out for this to come in utils 0.9.0 in the next wee while. If you need it before then, you can always install the package using git and pointing to the `next/minor` branch in the meantime.

**Comment by jpmmcneill** [comment]
brill, thank you Joel! 🤜 🤛

**Pull request merged**